lib/ai/model/tool.go

```diff
@@ -109,18 +134,281 @@ func (*commandExecutionTool) parseInput(input string) (*commandExecutionToolInpu
 	return &output, nil
 }
 
-type NodeGetter interface {
-	GetNodes(ctx context.Context, fn func(n services.Node) bool) []types.Server
-	NodeCount() int
+type accessRequestListRequestableRolesTool struct{}
+
+func (*accessRequestListRequestableRolesTool) Name() string {
+	return "List Requestable Roles"
+}
+
+func (*accessRequestListRequestableRolesTool) Description() string {
+	return "List all roles that can be requested via access requests."
+}
+
+func (a *accessRequestListRequestableRolesTool) Run(ctx context.Context, toolCtx *ToolContext, input string) (string, error) {
+	roles := toolCtx.AccessChecker.Roles()
+	requestable := make(map[string]struct{}, 0)
+	for _, role := range roles {
+		for _, requestableRole := range role.GetAccessRequestConditions(types.Allow).Roles {
+			requestable[requestableRole] = struct{}{}
+		}
+	}
+	for _, role := range roles {
+		for _, requestableRole := range role.GetAccessRequestConditions(types.Deny).Roles {
+			delete(requestable, requestableRole)
+		}
+	}
+
+	resp := strings.Builder{}
+	for role := range requestable {
+		resp.Write([]byte(role))
+		resp.Write([]byte("\n"))
+	}
+
+	if resp.Len() == 0 {
+		return "No requestable roles found", nil
+	}
+
+	return resp.String(), nil
+}
+
+type accessRequestListRequestableResourcesTool struct{}
+
+func (*accessRequestListRequestableResourcesTool) Name() string {
+	return "List Requestable Resources"
 }
 
-type embeddingRetrievalTool struct {
-	assistClient      assist.AssistEmbeddingServiceClient
-	nodeClient        NodeGetter
-	userAccessChecker services.AccessChecker
-	currentUser       string
+func (*accessRequestListRequestableResourcesTool) Description() string {
+	return `List all resources with IDs that can be requested via access requests.
+This includes nodes via SSH access.`
 }
 
+func (a *accessRequestListRequestableResourcesTool) Run(ctx context.Context, toolCtx *ToolContext, input string) (string, error) {
+	foundResources := make(chan promptResource, 0)
+	workersAlive := new(int32)
```

Comment:
You can use `atomic.Int32` here. Starting Go 1.19 AFAIR you don't need to play with pointers to use atomics anymore.